

浏览器是多线程的

+ GUI渲染线程

+ JS引擎线程

+ DOM监听线程/定时器监听线程

+ HTTP网络线程

+ ...



JS是单线程的（同时只能处理一件事情）

绝大部分是同步编程

有部分是单线程异步编程

浏览器基于EventQueue事件队列  和  EventLoop事件循环  构建出异步编程效果



+ 异步宏任务 macrotask
  + 定时器
  + DOM事件
  + 网络请求

+ 异步微任务 microtask
  + promise
  + async/await
  + requestAnimationFrame
  + ...



微任务队列中没有可执行的任务就去宏任务队列中查找，不是微任务队列为空才查找宏任务。

多个微任务都达到可执行状态，以达到执行状态的先后执行。



在WebAPI中监听异步任务，达到执行状态放入任务队列中

setTimeout 不放等待时间或者等待时间写0也不是同步立即执行的，而是根据浏览器默认的反应时间来执行，所以还是异步的。平常都说是4ms的间隔，但是最新的Chrome 浏览器是1 - 2ms。

JS相邻代码执行的间隔非常短，0.02ms左右





# PromiseA+ 规范

- 1.1. promise: 一个拥有符合这个规范的行为的`then`方法的对象或函数。
- 1.2. thenable: 定义了一个`then`方法的对象或函数。
- 1.3. 值(value): 任意合法的`JavaScript`值(包括`undefined`,`thenable`,`promise`)。
- 1.4. 异常(exception): 使用`throw`语句抛出的一个值
- 1.5. 原因(reason): 表示`promise`为什么被拒绝的一个值

### 2.1. Promise 状态

```
promise`必须是这三个状态中的一种：等待态`pending`,解决态`fulfilled`或拒绝态`rejected
```

- 2.1.1. 当一个`promise`处于等待状态的时候：

- - 2.1.1.1. 可能变为**解决**或者**拒绝**状态。

- 2.1.2. 当一个`promise`处于**解决**状态的时候：

- - 2.1.2.1. 一定不能转换为任何其它状态
  - 2.1.2.2. 必须有一个不能改变的**值**

- 2.1.3. 当一个`promise`处于**拒绝**状态的时候：

- - 2.1.3.1. 一定不能转换为任何其它状态
  - 2.1.3.2. 必须有一个不能改变的**值**

在这里，"一定不能改变"意味着不变的身份(例如 `===`)，但是并不意味着深度不可变性。(译注者：这里应该是说只要**值**的引用相同即可，并不需要引用中的每一个值都相等)

### 2.2. then 方法

`Promise`必须提供一个`then`方法来访问当前或最终的值或原因。

`Promise`的`then`方法接受俩个参数：

```js
promise.then(onFulfilled, onRejected)
```

### 2.2.1 `onFulfilled`和`onRejected`都是可选的参数

- 2.2.1.1. 如果`onFulfilled`不是一个函数，它必须被忽略
- 2.2.1.2. 如果`onRejected`不是一个函数，它必须被忽略

### 2.2.2. 如果`onFulfilled`是一个函数

- 2.2.2.1. 它必须在`promise`被**解决**后调用，`promise`的值作为它的第一个参数。
- 2.2.2.2. 它一定不能在`promise`被**解决**前调用。
- 2.2.2.3. 它一定不能被调用多次。

### 2.2.3. 如果`onRejected`是一个函数

- 2.2.3.1. 它必须在`promise`被**拒绝**之后调用，用`promise`的原因作为它的第一个参数。
- 2.2.3.2. 它一定不能在`promise`被**拒绝**之前调用。
- 2.2.3.3. 它一定不能被调用多次。

### 2.2.4. 在执行上下文栈中只包含平台代码之前，`onFulfilled`或`onRejected`一定不能被调用 [3.1]

### 2.2.5. `onFulfilled`和`onRejected`一定被作为函数调用(没有`this`值) [3.2]

### 2.2.6. 同一个`promise`上的`then`可能被调用多次

- 2.2.6.1. 如果`promise`被**解决**，所有相应的`onFulfilled`回调必须按照他们原始调用`then`的顺序执行
- 2.2.6.2. 如果`promise`被**拒绝**，所有相应的`onRejected`回调必须按照他们原始调用`then`的顺序执行

### 2.2.7. `then`必须返回一个`promise` [3.3]

```js
promise2 = promise1.then(onFulfilled,onRejected)
```

- 2.2.7.1. 如果`onFulfilled`或`onRjected`返回一个值`x`，运行`promise`解决程序`[[Resolve]](promise2,x)`
- 2.2.7.2. 如果`onFulfilled`或`onRejected`抛出一个异常`e`，`promise2`必须用`e`作为原因被**拒绝**
- 2.2.7.3. 如果`onFulfilled`不是一个函数并且`promise1`被**解决**，`promise2`必须用与`promise1`相同的值被**解决**
- 2.2.7.4. 如果`onRejected`不是一个函数并且`promise1`被**拒绝**，`promise2`必须用与`promise1`相同的原因被**拒绝**



# 源码

```js
(function () {
    'use strict'
    var PENDING = 'pending';
    var FULFILLED = 'fulfilled'
    var REJECTED = 'rejected';
    function Promise(executor) {
        var self = this, change;
        // 必须new执行 && executor必须是一个函数
        if (!(self instanceof Promise)) {
            throw new TypeError('undefined is not a promise');
        }
        if (typeof executor !== 'function') {
            throw new TypeError('Promise resolver ' + executor + ' is not a function');
        }

        // 实例具备的属性 状态&结果
        self.PromiseState = PENDING;
        self.PromiseResult = undefined;
        self.onFulfilledCallbacks = [];
        self.onRejectedCallbacks = [];
        change = function change(state, result) {
            if (state !== PENDING) {
                return;
            }
            self.PromiseState = state;
            self.PromiseResult = result;
            // 创建异步微任务，通知集合中方法执行
            var callbacks = state === FULFILLED ? self.onFulfilledCallbacks : self.onRejectedCallbacks,
                i = 0,
                len = callbacks.length,
                callback;
            if (callbacks.length) {
                setTimeout(() => {
                    for (; i < len; i++) {
                        callback = callbacks[i];
                        if (typeof callback === 'function') {
                            callback(result);
                        }
                    }
                });
            }
        };

        // executor 会立即执行
        try {
            executor(function resolve(result) {
                change(FULFILLED, result);
            }, function reject(reason) {
                change(REJECTED, reason);
            });
        } catch (error) {
            change(REJECTED, error);
        }
    }

    // 验证是否为promise实例
    function isPromise(x) {
        if (x == null) {
            return false;
        }
        if (/^(object|function)$/i.test(typeof x)) {
            if (typeof x.then === 'function') {
                return true;
            }
        }
        return false;
    }

    // 处理 方法执行返回值的处理
    function handle(promise, x, resolve, reject) {
        if (promise === x) {
            throw new TypeError('Chaining cycle detected for promise');
        }
        if (isPromise(x)) {
            try {
                x.then(resolve, reject);
            } catch (error) {
                reject(error);
            }
            return;
        }
        // 返回值不是promise实例，方法还没有报错，则是成功的
        resolve(x);
    }

    Promise.prototype = {
        constructor: Promise,
        self: true,
        then: function (onFulfilled, onRejected) {
            // 如果此时已经知道对应实例的状态是成功还是失败，则创建一个异步的微任务，后期执行对应的回调
            // 如果执行时还不知道状态，将两个回调存储起来，后期状态改变时通知执行。也是异步微任务
            // queueMicrotask 创建异步微任务 兼容性差 | 定时器创建 宏任务  模拟微任务

            if (typeof onRejected !== 'function') {
                onRejected = function onRejected(reason) {
                    throw reason;
                }
            }
            if (typeof onFulfilled !== 'function') {
                onFulfilled = function onFulfilled(result) {
                    return result;
                }
            }

            var x;
            var promiseNew = new Promise((resolve, reject) => {
                switch (this.PromiseState) {
                    case FULFILLED:
                        setTimeout(() => {
                            try {
                                x = onFulfilled(this.PromiseResult);
                                handle(promiseNew, x, resolve, reject);
                            } catch (error) {
                                reject(error);
                            }
                        });
                        break;
                    case REJECTED:
                        setTimeout(() => {
                            try {
                                x = onRejected(this.PromiseResult);
                                handle(promiseNew, x, resolve, reject);
                            } catch (error) {
                                reject(error);
                            }
                        });
                        break;
                    default:
                        this.onFulfilledCallbacks.push(function (result) {
                            try {
                                x = onFulfilled(result);
                                handle(promiseNew, x, resolve, reject);
                            } catch (error) {
                                reject(error);
                            }
                        });
                        this.onRejectedCallbacks.push(function (reason) {
                            try {
                                x = onRejected(reason);
                                handle(promiseNew, x, resolve, reject);
                            } catch (error) {
                                reject(error);
                            }
                        });
                }
            });
            return promiseNew;
        },
        catch: function (onRejected) {
            return this.then(null, onRejected);
        },
        finally: function (callback) {
            return this.then(function (result) {
                return Promise.resolve(callback()).then(()=>result)
            }, function (reason) {
                return Promise.resolve(callback()).then(()=>{throw reason})
            })
        }
    }

    if (typeof Symbol !== 'undefined') {
        Promise.prototype[Symbol.toStringTag] = 'Promise';
        // Object.prototype.toString.call() === '[object Promise]';
    }

    Promise.all = function all(promises) {
        if (!Array.isArray(promises)) {
            throw new TypeError(promises + 'is not iterable');
        }
        var promiseNew, results = [], count = 0;
        promises = promises.map((promise) => { // 控制每一项都是promise实例
            if (!isPromise(promise)) {
                return Promise.resolve(promise);
            }
            return promise;
        });
        promiseNew = new Promise(function (resolve, reject) {

            promises.forEach(function (promise, index) {
                promise.then(function (result) {
                    results[index] = result;
                    count++;
                    if (count >= promises.length) {
                        resolve(results);
                    }
                }).catch(function (reason) {
                    reject(reason);
                });
            })
        });
        return promiseNew;
    }

    Promise.resolve = function resolve(result) {
        return new Promise(function (resolve) {
            resolve(result);
        });
    }

    Promise.reject = function reject(reason) {
        return new Promise(function (_, reject) {
            reject(reason);
        });
    }

    if (typeof window !== 'undefined') {
        window.Promise = Promise;
    }
    if (typeof module === 'object' && (typeof exports === 'object')) {
        module.exports = Promise;
    }
})()

```



# Iterator

遍历器是一种机制，为各种不同的数据结构提供统一的访问机制，任何数据结构只要部署iterator接口，就可以完成遍历操作（for  of），依次处理该数据结构的所有成员。

拥有next方法用于依次遍历成员。

每次遍历返回的结果是一个对象

done：记录是否遍历完成。

value：当前遍历的结果。



所有具有Symbol.iterator属性的数据结构，被称为可遍历的。for of 循环根据iterator规范进行遍历。



# genetator 生成器



```js
function* fn() {
  
}
let gen = fn(); // 函数调用 但是函数体的代码并没有执行
console.log(gen) // 返回的一个具有迭代器规范的对象 gen相当于fn的一个实例
// gen.__proto__ => Generator.prototype => GeneratorFunction.prototype(next/return/throw) => xxx.prototype(Symblo.iterator) => Object.prototype

gen.next(); // 开始执行函数体代码 返回{value: xxx, done: true/false} 对象
gen.return(); // 结束遍历，done变为true，参数为value值
gen.throw(); // 直接抛出异常，结束遍历，后面的next 无法执行
```

